core/templates/dev/head/domain/statistics/PlaythroughImprovementCardObjectFactory.js

```diff
@@ -0,0 +1,232 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory for creating Playthrough Cards in the Improvements Tab.
+ *
+ * NOTE: For testing and organizational purposes, improvement card object
+ * factories should be registered in the ImprovementCardRegistryService! See
+ * the corresponding file to see what functions are expected from cards, and how
+ * to add new ones as necessary.
+ */
+
+oppia.factory('PlaythroughImprovementCardObjectFactory', [
+  '$uibModal', 'ImprovementActionObjectFactory', 'PlaythroughIssuesService',
+  'UrlInterpolationService',
+  function(
+      $uibModal, ImprovementActionObjectFactory, PlaythroughIssuesService,
+      UrlInterpolationService) {
+    var openPlaythroughModal = function(playthroughId, index) {
+      PlaythroughIssuesService.getPlaythrough(playthroughId).then(
+        function(playthrough) {
+          $uibModal.open({
+            templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
+              '/pages/exploration_editor/statistics_tab/' +
+              'playthrough_modal_directive.html'),
+            backdrop: true,
+            resolve: {
+              playthrough: function() {
+                return playthrough;
+              },
+              playthroughIndex: function() {
+                return index;
+              }
+            },
+            controller: [
+              '$scope', '$uibModalInstance', 'playthroughIndex',
+              'playthrough', 'AlertsService', 'LearnerActionRenderService',
+              function(
+                  $scope, $uibModalInstance, playthroughIndex,
+                  playthrough, AlertsService, LearnerActionRenderService) {
+                $scope.playthroughIndex = playthroughIndex;
+
+                $scope.displayBlocks =
+                  LearnerActionRenderService.getDisplayBlocks(
+                    playthrough.actions);
+                $scope.reversedDisplayBlocks =
+                  $scope.displayBlocks.slice().reverse();
+
+                var blockActionIndexMapping = {};
+                $scope.displayBlocks.reduce(
+                  function(runningTotal, displayBlock, i) {
+                    blockActionIndexMapping[i] =
+                      runningTotal - displayBlock.length;
+                    return blockActionIndexMapping[i];
+                  }, playthrough.actions.length + 1);
+
+                $scope.maxHidden = $scope.displayBlocks.length - 1;
```

Comment:
Hmm, a lot of this code can be improved in general, but I don't want to include that in this PR since it's a bit unrelated (nothing broken). For now, this PR will only focus on dedupeing usages of this code (opening the modal).